[1051] Add documentation for configuration of smartrest templates (#1147)
thin-edge/thin-edge.io: docs/src/howto-guides/023_smartrest_templates.md

# How to add C8Y SmartRest Templates

[SmartRest Templates](https://cumulocity.com/guides/10.6.0/reference/smartrest/#templates) can be used to add custom templates to the SmartRest protocol.

thin-edge.io supports topic subscription to custom templates as per [C8Y documentation](https://cumulocity.com/guides/10.6.6/users-guide/device-management/#smartrest-templates).

> All configured templates are going to generate following MQTT subscriptions:
> 
> ```plain
> s/dc/template-name
> s/uc/template-name
> ```

Reviewer: ```suggestion
A template named `$TEMPLATE_NAME` is requires the following subscriptions to be added when connecting Cumulocity:

```plain
s/dc/$TEMPLATE_NAME
s/uc/$TEMPLATE_NAME
``

__This is not done automatically and the custom templates have to be declared using the `tedge` command__.
````

> ```

Reviewer 2: This suggestion might still be relevant as it mentions the local bridged topics that the end-user need to care about.

## Checking existing templates

```shell
tedge config get c8y.smartrest.templates

```

## Add new template to thin-edge configuration

To add new template to thin-edge.io a tedge cli tool can be used:

```shell
tedge config set c8y.smartrest.templates template-1,template-2
```

> > Note: If you adding templates to already existing ones you have to add all of the template together in one command:
> > For example we have following template `template-1` already in our configuration

Reviewer 2: ```suggestion
> Note: To add/append a new template to a device that's already configured with some, all the existing templates should also be declared along with the new one in the `tedge config set` command.
> For example, if `template-1` is already configured on the device, as follows:
```

> # How to use C8Y Custom SmartRest 2 Templates

Reviewer 3: To be more precise,
```suggestion
# How to use Cumulocity Custom SmartREST 2.0 Templates
```

[Custom SmartRest Templates](https://cumulocity.com/guides/reference/smartrest-two) can be used to extend the functionality of a device to support more operations than what the [static SmartREST templates](https://cumulocity.com/guides/reference/smartrest-two/#mqtt-static-templates) offer.

thin-edge.io supports subscription to custom templates as documented [here](https://cumulocity.com/guides/users-guide/device-management/#smartrest-templates).

For every template that the device uses, it must publish all data to `s/uc/<template-name>` topic and subscribe to `s/dc/<template-name>` to receive data from the cloud, based on that template.

> When these templates are configured with thin-edge.io, subscriptions to all these relevant topics on Cumulocity cloud will be done by thin-edge internally. Local processes on the device can access these templates on the local MQTT broker by simply publishing to `c8y/s/uc/<template-name>` and subscribing to `c8y/s/dc/<template-name>` topics (note the `c8y/` prefix in topics).

Reviewer 3: To keep consistency of `thin-edge.io`. 
```suggestion
When these templates are configured with thin-edge.io, subscriptions to all these relevant topics on Cumulocity cloud will be done by thin-edge.io internally. Local processes on the device can access these templates on the local MQTT broker by simply publishing to `c8y/s/uc/<template-name>` and subscribing to `c8y/s/dc/<template-name>` topics (note the `c8y/` prefix in topics).
```

A template named `$TEMPLATE_NAME` requires the following subscriptions to be added when connecting to Cumulocity:

````plain
s/dc/$TEMPLATE_NAME

> s/uc/$TEMPLATE_NAME

Reviewer 2: I don't think a subscription to this `s/uc` topic is needed. `s/uc` is used only for publishing data. `s/dc` is the one that must be subscribed to.

> ``

Reviewer 3: The format of this block seems broken.
Line 12 has four backticks and line 15 has two.

__This is not done automatically and the custom templates have to be declared using the `tedge` command.__

## Checking existing templates

```shell
tedge config get c8y.smartrest.templates

> ````

Reviewer 3: Also having 4 backticks instead of 3.
